library/src/tests/diagnostics.rs

```diff
@@ -40,3 +40,119 @@ fn check_operations_are_equal() {
         ),
     );
 }
+
+#[test]
+fn check_start_stop_counting_operation_called_3_times() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingOperation;
+            import Microsoft.Quantum.Diagnostics.StopCountingOperation;
+
+            operation op1() : Unit {}
+            operation op2() : Unit { op1(); }
+            StartCountingOperation(op1);
+            StartCountingOperation(op2);
+            op1(); op1(); op2();
+            (StopCountingOperation(op1), StopCountingOperation(op2))
+        }",
+        &Value::Tuple([Value::Int(3), Value::Int(1)].into()),
+    );
+}
+
+#[test]
+fn check_start_stop_counting_operation_called_0_times() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingOperation;
+            import Microsoft.Quantum.Diagnostics.StopCountingOperation;
+
+            operation op1() : Unit {}
+            operation op2() : Unit { op1(); }
+            StartCountingOperation(op1);
+            StartCountingOperation(op2);
+            (StopCountingOperation(op1), StopCountingOperation(op2))
+        }",
+        &Value::Tuple([Value::Int(0), Value::Int(0)].into()),
+    );
+}
+
+#[test]
+fn check_stop_counting_operation_without_start() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StopCountingOperation;
+
+            operation op1() : Unit {}
+            StopCountingOperation(op1)
+        }",
+        &Value::Int(-1),
+    );
+}
+
+#[test]
+fn check_counting_operation_differentiates_between_body_adj_ctl() {
+    test_expression(
+        "{
+            import Microsoft.Quantum.Diagnostics.StartCountingOperation;
+            import Microsoft.Quantum.Diagnostics.StopCountingOperation;
+
+            operation op1() : Unit is Adj + Ctl {}
+            StartCountingOperation(op1);
+            StartCountingOperation(Adjoint op1);
+            StartCountingOperation(Controlled op1);
+            op1();
+            Adjoint op1(); Adjoint op1();
+            Controlled op1([], ()); Controlled op1([], ()); Controlled op1([], ());
+            (StopCountingOperation(op1), StopCountingOperation(Adjoint op1), StopCountingOperation(Controlled op1))
+        }",
+        &Value::Tuple([Value::Int(1), Value::Int(2), Value::Int(3)].into()),
+    );
+}
+
```

Comment:
In the chat we discussed the subtleties of tracking if we define "let my_x = X" and track my_x, vs if we define "let my_x = q => X(q)" and track my_x. I'll need a test case for this

Reply:
And mentioning this in API docs would be helpful, I'm bound to forget it otherwise

Reply:
Added a test case. I'll be doing some big updates to the doc comments to cover all the feedback.